dilated-file-transfer.md

```diff
@@ -0,0 +1,484 @@
+This allows for existing messages to be extended, or for new message types to be added.
+Peers MUST _accept_ messages for any features they declare in `"features"`.
+Peers MUST only send messages / attributes for features in the other side's list.
+Only one format exists currently: `"core"`.
+
+   XXX:: maybe just lean on "version" for now? e.g. version `2` could introduce "features"?
+
+See "Example of Protocol Expansion" below for discussion about adding new attributes (including when we might increment the `"version"` instead of adding a new `"feature"`).
+
+The `"permission"` key specifies how to proceed with offers.
+Using mode `"ask"` tells the sender to pause after transmitting the first metadata message and await an answer from the peer before streaming further data.
+Using mode `"yes"` means the peer will accept all incoming transfers so the sender should not pause after the metadata (and instead immediately start sending data messages).
+Although a peer could implement "yes" mode by simply sending an accept message for each offer without user interaction, setting this mode cuts down latency for "one-way" transfers (see [Discussion](#discussion))
+
+
+## Protocol Details
+
+See the Dilation document for details on the Dilation setup procedure.
+Once a Dilation-supporting connection is open, we will have a "control" subchannel (subchannel #0).
+Either peer can also open additional subchannels.
+
+All control-channel messages are encoded using `msgpack`.
+   --> XXX: see "message encoding" discussion
+
+Control-channel message formats are described using Python pseudo-code to illustrate the data types involved.
+They are actually an encoded `Map` with `String` keys (to use `msgpack` lingo) and values as per the pseudo-code.
+
+All control-channel messages contain a integer "kind" field describing the sort of message it is.
+(That is, `"kind": 1` for example, not the single-byte tag used for subchannel messages)
+
+Rejected idea: Version message, because we already do version negotiation via mailbox features.
+
+Rejected idea: Offer/Answer messages via the control channel: we need to open a subchannel anyway and the subchannel-IDs are not intended to be part of the Dilation public API.
+
+
+### Control Channel Messages
+
+Each side MAY send a free-form text message at any time.
+These messages look like:
+
+```python
+class Message:
+    message: str     # unicode string
+    kind: int = 1    # "text message"
+```
+
+
+### Making an Offer
+
+Either side MAY propose any number of Offers at any time after the connection is set up.
+If the other peer specified `"mode": "send"` then this peer MUST NOT make any Offers.
+
+To make an Offer the peer opens a subchannel.
+Recall from the Dilation specification that subchannels are _record_ pipes (not simple byte-streams).
+
+All records on the subchannel begin with a single byte indicating the kind of message.
```

Comment:
Why not just use msgpack for encoding the message all the way down? This would work better with non-numeric message type identifiers anyways.

Reply:
Then we have to run all the data bytes through the msgpack parser. This would then imply choosing a "no-parse" format instead of msgpack.

Reply:
Is parsing really a (performance) concern here? I don't think it makes a big difference: In any case, the data is just laid out as a flat byte array somewhere in memory. Performance aware implementations may choose to be zero-copy and just have a pointer to the bytes, others that don't case will just memcpy the bytes to somewhere else (since we already have decryption going the data is already copied around multiple times so I don't think it matters).

Ballpark estimate: Even at a gigabit transfer rate and 4kiB packet size, there will only be ~30,000 packets per second, leaving an order of magnitude of 100,000 CPU cycles for parsing per packet on a single-threaded gigahertz processor. Parsing won't be our bottleneck any time soon.

Reply:
I haven't measured msgpack (or other parsers) performance, but the whole point of things like flatbuffers and cap-n-proto existing at all is to do the "pointer to bytes" trick you suggest (and I am not aware of any msgpack library that does that).

In Autobahn profiles, at least 60% of CPU time went to parsing. This is indeed a different situation, but I don't think we should completely ignore it unless there's a very compelling reason.

Reply:
The "pointer to bytes" trick is trivial for byte fields and works even for JSON. What those formats do is apply the same concept to entire *structures*, which requires a lot more control over the memory layout of those structures.

I won't die on this hill, but can we please at least have four bytes of ID space just to be *extra* sure?

Reply:
Another option is to just depend on the position of the records to indicate what they are. Everything that _isn't_ file-payload is msgpack-encdoed. First message is offer. Any reply is reject or accept. Rest are payloads.

For directories, first message is offer, second message is file-offer, then enough payloads to full the bytes indicated, then a file-offer, then more payloads, etc.

Reply:
I'm always in favor of more linear control flow (results in less state machine states), but I fear that as soon as we account for cancellation and error handling that approach goes out of the window.

The classic file transfer has this issue, where it just sends over bytes and at the end an ack, but if the file somehow contains less (or more) bytes than advertised, everything goes horribly wrong.

Reply:
I don't think any of these options affect the state-machine.

Consider receiving an unexpected message: it doesn't matter if that's because it didn't msgpack-decode, or the "kind" tag was wrong, or an unexpected "kind" string -- it's still an error. I do _feel_ like it's more fragile to rely on simply position (but in the end, the important feature is if you get to "error" state when you receive the wrong message at the wrong time, no matter how it's encoded). 

If the file has the wrong number of bytes, it's an error.

Essentially the error-handling / cancellation here is "close the subchannel". If you do that before reaching the end of the protocol, the transfer has failed.

How does the original transfer protocol get into a situation where it reaches the "ack" but doesn't have the right number of bytes? Have you seen that?

Anyway, if using tags we could certainly add more bytes to the tag, but also we'd have to introduce at least 250 more different _kinds_ of messages before it's a problem -- and so some future revision/feature could do that? (i.e. the feature that needs 255th new message could introduce 0xff to mean "look at the next byte" etc). Indeed, this is kind of what msgpack does for bytestrings, ints etc (there's different type tags for longer ones) and it uses a single leading byte for type (and other) information.

Reply:
> How does the original transfer protocol get into a situation where it reaches the "ack" but doesn't have the right number of bytes? Have you seen that?

Scenario 1: If the file got appended to during the transfer, it will read some random part of the file as "ack" message. This is okay, because it will very likely fail verification.

Scenario 2: The file got truncated somewhere during the transfer, or the actual length is shorter than expected in some other way. Then the ack message will be written out as part of the file, and the receiver will hang forever waiting for the remaining bytes, which will obviously never come.

Reply:
oh, right, filesystems change .. I was thinking one side sent more than the other read etc.